op-batcher/batcher/channel_manager.go

```diff
@@ -195,8 +203,11 @@ func (s *channelManager) TxData(l1Head eth.BlockID) (txData, error) {
 		return txData{}, err
 	}
 
-	if err := s.processBlocks(); err != nil {
-		return txData{}, err
+	// Avoid processing blocks if the channel manager has been explicitly closed.
+	if !s.closed {
+		if err := s.processBlocks(); err != nil {
```

Comment:
Note: I think this makes sense, but `s.processBlocks()` would return anyways without adding any blocks because the channel is marked as full.

Reply:
This is actually currently necessary in order to handle the case where the `pendingChannel` is `nil`. More likely we should avoid creating a new `pendingChannel` (via `ensurePendingChannel`) whenever this channel manager is already closed.

Reply:
You're right, I agree that it probably makes sense to not even call `ensurePendingChannel` if `s.closed`. To avoid moving more and more stuff into the `if !s.closed {...}` block, we could extend above early exit to
```
	// Short circuit if there is a pending frame or we're shutting down
	if dataPending || s.closed {
		return s.nextTxData()
	}
```

For that to work, we would need to call `s.outputFrames()` at the end of `channelManager.Close()`, to call it one last time when shutting down. (Did we have that in an earlier iteration? 😅 )

Reply:
Haha I think that's where we started: https://github.com/ethereum-optimism/optimism/pull/5105/commits/a9318e6fc89a221a02b1d0760133cd545e108242#r1142729550

We chose to take this approach in order to avoid double-closing the channel builder, but I'll see if that can be avoided.